### PR TITLE
Fixed an issue caused by escaping dollar signs.

### PR DESCRIPTION
**Pull request by daiaji**
Fix the faults mentioned below.

https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/issues/3239

### CODE REVIEW

**Comment by daiaji** [comment]
起码shell脚本的$1、$2变量不会被瞎转义了。

**Comment by H0llyW00dzZ** [comment]
does it will be fine for latex ?

**Comment by H0llyW00dzZ** [comment]
> > does it will be fine for latex ?
> 
> This formula is valid, basically just adding a judgment that the dollars in the code block will not be escaped.
> 
> But this is basically the solution given by GPT3.5, and I would like someone to review the code for me.


Do test using these masks:

[Exploring Polynomial Equations: From Quadratics to Cubics [LaTex Language].json](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/files/14398885/Exploring.Polynomial.Equations_.From.Quadratics.to.Cubics.LaTex.Language.json)

You can simply change the model according to the example format provided below:

Example format:
`Find the roots of the equation: $$2x^2 - 5x + 2 = 0$$`


this example how passed:

image

**Comment by H0llyW00dzZ** [comment]
@daiaji your deployment `https://4124-vercelpreviewdeployment.pr.nextchat.dev/`  it just break `$$` which is indicate that latex

image

**Comment by daiaji** [comment]
> @daiaji your deployment `https://4124-vercelpreviewdeployment.pr.nextchat.dev/` it just break `$$` which indicate that latex
> 
> image

Basically I just added a judgment that the dollar sign escape function does not escape the dollar sign in the code block, no other fixes, but I can have GPT3.5 try to see if I can fix this bug.

**Comment by H0llyW00dzZ** [comment]
> > @daiaji your deployment `https://4124-vercelpreviewdeployment.pr.nextchat.dev/` it just break `$$` which indicate that latex
> > image
> 
> Basically I just added a judgment that the dollar sign escape function does not escape the dollar sign in the code block, no other fixes, but I can have GPT3.5 try to see if I can fix this bug.

you may try to fix, because this one it fix for latex

https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/blob/e756506c185dcd786ad04db0c7f2c423a6496072/app/components/markdown.tsx#L102

File: app/components/markdown.tsx [hunk 1]
```diff
@@ -100,17 +100,33 @@ export function PreCode(props: { children: any }) {
...
 
-    escapedText += char;
+    if (char === "`") {
+      inCodeBlock = !inCodeBlock;
+      escapedText += "`";
+      continue;
+    }
+
+    // Check for $ followed by a number
+    if (!inCodeBlock && char === "$" && /^[0-9]$/.test(nextChar)) {
```
Comment (H0llyW00dzZ) [inline app/components/markdown.tsx:125]: this one it will break `$$` for LaTeX

**Review by H0llyW00dzZ — COMMENTED** [review]
> To be honest, just describing the fault performance to GPT is very troublesome, and GPT3.5's poor logical ability is exposed at this time, and GPT3.5 cannot even understand my statement. I am trying GPT4.

The model is not the issue. The key to correct LaTeX formatting is the use of `$` and `$` for inline and `$$` and `$$` for block expressions. So, if you receive output from the AI that doesn't include these for LaTeX, it's an AI issue, possibly related to the prompt. This is because `$` and `$$` are the correct LaTeX syntax for Markdown format. The problem is detailed below:

**Comment by daiaji** [comment]
Good news, GPT has provided usable code. In my opinion, the code logic is too complex, but at least the code can work. Now I hope GPT can optimize the code, and I think I can upload the code soon.😊

File: app/components/markdown.tsx (hunk 1 above)
Comment (daiaji) [reply app/components/markdown.tsx:125]: 新的聊天
The rendering seems to have been corrected!

**Comment by H0llyW00dzZ** [comment]
> Good news, GPT has provided usable code. In my opinion, the code logic is too complex, but at least the code can work. Now I hope GPT can optimize the code, and I think I can upload the code soon.😊

does it will be fine for latex ?

**Comment by daiaji** [comment]
> does it will be fine for latex ?

The example you provided seems to have been rendered correctly, and I found another formula that also appears to have been rendered correctly. The $in the code block also seems to have been correctly escaped.

To be honest, it feels like dealing with this kind of problem is logically cumbersome and may still require code review.🧐

**Comment by H0llyW00dzZ** [comment]
> > does it will be fine for latex ?
> 
> The example you provided seems to have been rendered correctly, and I found another formula that also appears to have been rendered correctly. The $in the code block also seems to have been correctly escaped.
> 
> To be honest, it feels like dealing with this kind of problem is logically cumbersome and may still require code review.🧐

@daiaji now this one causing bug

image

anyway it's difficult to maintain/fix cause of complexity, even I was trying it before

**Comment by daiaji** [comment]
> anyway it's difficult to maintain/fix cause of complexity, even I was trying it before

No surprise, I always feel like it will become very troublesome in the end.

**Comment by daiaji** [comment]
> Yes, because of its complexity, it's quite difficult to maintain or fix. In contrast, breaking down each function into smaller parts to make it modular makes maintenance much easier.

I think irresponsibly, if the complexity of the code increases infinitely, save GPT to maintain as much modularity as possible!🤣

**Comment by H0llyW00dzZ** [comment]
> > Yes, because of its complexity, it's quite difficult to maintain or fix. In contrast, breaking down each function into smaller parts to make it modular makes maintenance much easier.
> 
> I think irresponsibly, if the complexity of the code increases infinitely and the code has been modularized as much as possible, let GPT maintain the code!🤣

429543792_809180301237785_3392338544743453221_n

**Comment by daiaji** [comment]
> _**It was as if beholding a vista not even the most horrifying of nightmares could conjure, an indescribable maelstrom of thoughts entwined around my very soul. Perhaps, this was the answer bestowed by those blasphemous entities, older than time itself.**_

https://sharegpt.com/c/WJbLuB8
新的聊天

**Comment by H0llyW00dzZ** [comment]
> > _**It was as if beholding a vista not even the most horrifying of nightmares could conjure, an indescribable maelstrom of thoughts entwined around my very soul. Perhaps, this was the answer bestowed by those blasphemous entities, older than time itself.**_
> 
> https://sharegpt.com/c/WJbLuB8 新的聊天

@daiaji

image

image

**Comment by H0llyW00dzZ** [comment]
bug its on single backticks now

**Comment by H0llyW00dzZ** [comment]
@daiaji it's difficult to fix

image

this pattern causing bug:

`$1 $1 $1 $1`

$1 $2 $3 $4

`$1 $2 $3 $4`

`$1 $2 $3 $4`

---

$1 $2 $3 $4



image

**Comment by H0llyW00dzZ** [comment]
another bug:

image

**Comment by H0llyW00dzZ** [comment]
Anyway There appears to be an issue, possibly between React Markdown and the UI. Although the code is correct, the UI or React Markdown seems to be causing some unexpected behavior.

**Comment by daiaji** [comment]
新的聊天(1)

**Comment by daiaji** [comment]
It seems that the complexity of this function group is already very high.

**Comment by H0llyW00dzZ** [comment]
> Although this function is already quite complex, it may still require expanding the scope of the check.

Yeah its difficult to fix/maintain about this front end, unlike backend written in go

**Comment by H0llyW00dzZ** [comment]
Also It's not possible maybe to fix because the function is already reached complexity which is not recommended

**Comment by daiaji** [comment]
I can only hope that this complexity stops there.

**Comment by daiaji** [comment]
It's better to merge the code first than not fix the bug at all.

**Comment by daiaji** [comment]
https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/issues/3860 Does this meet expectations?

**Comment by H0llyW00dzZ** [comment]
> It's better to merge the code first than not fix the bug at all.

I am not owner, the owner is @fred-bf 

> #3860 Does this meet expectations? Or is this normal?

no it's bug

**Comment by daiaji** [comment]
> no it's bug

Maybe it's a glitch in markdown's rendering library, maybe create an input preprocessor that looks like a dollar sign? Does that sound terrible?

**Comment by H0llyW00dzZ** [comment]
> > no it's bug
> 
> Maybe it's a glitch in markdown's rendering library, maybe create an input preprocessor that looks like a dollar sign? This sounds terrible, and the escape symbol is 10,000 times more common than the dollar sign!

no glitch, it literally bug even in github markdown

([a-z])\([a-z])
`([a-z])\([a-z])`

**Comment by H0llyW00dzZ** [comment]
expected should be same inside backtick and outside backticks but it's different lmao

**Comment by H0llyW00dzZ** [comment]
google ai studio works fine that pattern

image

**Comment by daiaji** [comment]
> 谷歌人工智能工作室在这种模式下工作得很好
> 
> 图像

Maybe it's just a tendency problem, and to be honest, even if it's a bug, it's a few dozen times more difficult to fix than the dollar sign.

**Comment by daiaji** [comment]
To be honest, I want to close this PR because the incorrect link is causing me discomfort.🙁

**Comment by H0llyW00dzZ** [comment]
> To be honest, I want to close this PR because the incorrect link is causing me discomfort.🙁

you can close it, its you pr

**Pull request closed without merging**